Simplify logic and flow of ParentCasePropertyBuilder

dannyroberts opened the pull request:
Fixes https://manage.dimagi.com/default.asp?269202.

Includes a BIG refactor of the underlying code, going from a messy recursive process, to a (I like to think) streamlined process based on aggregating inputs from all possible sources (forms in apps, data dictionary, etc.) in the same format (`{<case_type>: set([<contributed case property>])`), flattening all of that information (which includes information for _all_ case types), and then doing a one-pass (or fixed-N-pass depending on when you start counting) process to propagate all the parent/child refs up and down. The result is a process that's more easily debuggable and easier to reason with.

Other than fixing the long-standing bug, another nice thing about this new organization is that it is very easy to (in the future) isolate the contribution coming form each app and quickcache that, thus having to pull only (1) the full list of app_ids to include and (2) look up the case property contributions of that app in the cache. Previously, the code kept coming back to the app to look for more things at different times in the recursive process; now it's all pulled up-front and processed separately.

This PR also makes it obvious that when you ask for all of the properties for a single case type, you actually basically have to look up all properties for _all_ case types to make sure you're getting parent child properties correctly propagated. (While technically it's true that you could have separate hierarchically unrelated case types all of which are now getting pulled in, where maybe (maybe?) before they weren't, in my estimation, the overhead here is _by far_ pulling all the apps, which still had to be done before.) The reason I say now it's obvious is that now all of the actual work is done by pulling case contributions for all case types up front, and the methods for getting stuff for a single case type just call those methods and filter down to a single case type. (Things are nicely memoized, so it's not actually less efficient to call these per-case-type methods for multiple different case types.)

dannyroberts committed:
remove no-longer-true comment

dannyroberts committed:
pull out get_case_relationships

to remove a direct use of ParentCasePropertyBuilder

dannyroberts committed:
mark private ParentCasePropertyBuilder methods

dannyroberts committed:
pull out _get_forms_info

dannyroberts committed:
pull out _get_all_forms_info

dannyroberts committed:
pull out get_all_case_updates in Form

to get all case updates rather than pre-filtering by case type

dannyroberts committed:
pull out get_all_case_updates in AdvancedForm

just like was just done in Form

dannyroberts committed:
rename get_case_updates => get_case_updates_by_case_type

dannyroberts committed:
add ParentCasePropertyBuilder._get_all_case_updates

to collate all calls to form.get_all_case_updates

dannyroberts committed:
break form.get_parent_types_and_contributed_properties up

into constituent parts:
- get_contributed_subcase_properties
- get_contributed_parent_types

dannyroberts committed:
break form.get_parent_types_and_contributed_properties up (AdvancedForm)

dannyroberts committed:
break ParentCasePropertyBuilder.get_parent_types_and_contributed_properties up

dannyroberts committed:
remove all get_parent_types_and_contributed_properties methods

replace their usages with the appropiate constituent method

dannyroberts committed:
scrap tracking module type for forms in ParentCasePropertyBuilder

dannyroberts committed:
add case properties test for case sharing

make sure that apps get each others's case properties.
Test fails when these lines
https://github.com/dimagi/commcare-hq/blob/85321a01df52a59f7c3345d9320bca821bd3359b/corehq/apps/app_manager/app_schemas/case_properties.py#L130-L138
are commented out

dannyroberts committed:
add case properties test for case sharing

make sure that apps get each others's case properties.
Test fails when these lines
https://github.com/dimagi/commcare-hq/blob/85321a01df52a59f7c3345d9320bca821bd3359b/corehq/apps/app_manager/app_schemas/case_properties.py#L130-L138
are commented out

dannyroberts committed:
get case sharing properties much more simply

dannyroberts committed:
rename _get_all_forms => _get_relevant_forms

and clean up helper methods

dannyroberts committed:
rename "parent types" to "case relationships"

when it is actually (case_type, relationship_type)

dannyroberts committed:
make get_all_contributed_subcase_properties not filter on case type

but rather calculate for all case types at once

dannyroberts committed:
push get_all_contributed_subcase_properties into get_all_case_updates

now rather than fetching the contrubution of forms opening subcases as a separate step
ParentCasePropertyBuilder receives these as well from _get_case_updates

dannyroberts committed:
rearrange how/where include_parent_properties is implemented

dannyroberts committed:
rename to get_case_updates_for_case_type for less ambiguity

dannyroberts committed:
make sure get_case_updates_for_case_type has its former behavior

because it is called strangely from export and I don't
want to have to figure that out right now

dannyroberts committed:
move get_save_to_case_updates into get_all_case_updates too

dannyroberts committed:
use ParentCasePropertyBuilder._get_all_case_updates

rather than the version pre-filtered by case type

dannyroberts committed:
make get_case_relationships do all case types

dannyroberts committed:
clean up get_case_relationships_for_case_type use in get_properties

dannyroberts committed:
simplify use of updates_by_case_type in get_properties

millerdev commented on corehq/apps/app_manager/app_schemas/case_properties.py:
```diff
@@ -18,117 +18,106 @@
+        apps = [self.app]
+        if self.app.case_sharing and include_shared_properties:
+            for app in self._get_other_case_sharing_apps_in_domain():
+                apps.append(app)
```
nit: these two lines could be reduced to `apps.extend(self._get_other_case_sharing_apps_in_domain())`

millerdev commented on corehq/apps/app_manager/app_schemas/case_properties.py:
```diff
@@ -18,117 +18,106 @@
+    forms = []
+    for module in app.get_modules():
+        for form in module.get_forms():
+            forms.append(form)
```
nit: the above two lines could be reduced to `forms.extend(module.get_forms())`

dannyroberts committed:
pull out _get_data_dictionary_properties_by_case_type

millerdev commented on corehq/apps/app_manager/app_schemas/case_properties.py:
```diff
@@ -18,117 +19,118 @@
+        return self.get_case_relationships(include_shared_properties)[case_type]
+
+    @memoized
+    def get_case_relationships(self, include_shared_properties=True):
```
Would be nice for all (at least non-private) functions like this to have a docstring that describes the data structure of the returned object. For example, this one might be
```
Returns a `defaultdict(set)` where each key is a case type
(string) and each value is a set of ...
```
I'm not even sure how to finish that, but guessing it's a set of relationship names (strings) like `"parent"`.

millerdev reviewed (commented):
I know you're not finished yet. I added a few, mostly nitpicky, comments.

dannyroberts committed:
prefer extend to append in a loop

dannyroberts committed:
fix indent

dannyroberts committed:
continue to build all properties by type all together

dannyroberts committed:
simplify get_parent_type_map

dannyroberts committed:
recognize parent/foo on the child as a property on the parent

- make get_properties a shortcut for get_properties_by_case_type(...)[case_type]
- replace recursive call to get_properties with queue-based iterative process
  that is pulled out separately

dannyroberts committed:
uncomment now-fixed test

See commit where it was first added and noted:
https://github.com/dimagi/commcare-hq/pull/12319/files#diff-b9de4a695c074e8e8121b6e74f402a9dR168

Used to fail with:
```
======================================================================
ERROR: corehq.apps.app_manager.app_schemas.tests.test_schema:SchemaTest.test_get_casedb_schema_with_parent_case_property_update
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/droberts/.virtualenvs/cchq/lib/python2.7/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/Users/droberts/dimagi/commcare-hq/corehq/apps/app_manager/app_schemas/tests/test_schema.py", line 125, in test_get_casedb_schema_with_parent_case_property_update
    self.assertEqual(subsets["parent"]["structure"]["has_well"], {"description": ""})
KeyError: u'has_well'

----------------------------------------------------------------------
```

dannyroberts committed:
remove now-unused include_shared_properties

dannyroberts committed:
make _get_usercase_default_properties private

dannyroberts committed:
move include_parent_properties to ParentCasePropertyBuilder

dannyroberts committed:
add test for parent/child property behavior

dannyroberts committed:
Merge branch 'dmr/case-properties-tests' into dmr/case-properties

dannyroberts committed:
fix test_parent_child_properties

and uncomment the now-fixed lines!

dannyroberts committed:
import future print_function in case_properties.py

dannyroberts committed:
fix simple bug in dictionary properties refactor

dannyroberts committed:
fix stickler warning

about USERCASE_TYPE being from a star import

dannyroberts committed:
use namedtuples where possible

dannyroberts committed:
continue to use more explicit name instead of `reference`

for more consistent naming:
- equivalence
- case_type_ref
- expansion (also a case_type_ref, but one that is thought of as an expansion of a simple case_type)
- property_ref

dannyroberts committed:
document new approach to propagating case properties

and lightly test with a doctest

dannyroberts committed:
more fully document ParentCasePropertyBuilder

dannyroberts committed:
add implementation note describing new one-pass process for normalizing case properties

dannyroberts committed:
make _CaseRelationshipManager doctest more readable

dannyroberts commented on corehq/apps/app_manager/app_schemas/tests/test_schema.py:
```diff
@@ -122,7 +122,7 @@ def test_get_casedb_schema_with_parent_case_property_update(self):
         }})
         self.assertEqual(subsets["case"]["structure"]["case_name"], {"description": ""})
-        #self.assertEqual(subsets["parent"]["structure"]["has_well"], {}) TODO
+        self.assertEqual(subsets["parent"]["structure"]["has_well"], {"description": ""})
```
Finally getting to uncomment this line!
https://github.com/dimagi/commcare-hq/commit/33642a7bbf942fb17bbfa652e72c54f16bd0f189#diff-b9de4a695c074e8e8121b6e74f402a9dR168

millerdev replied:
🎉

dannyroberts commented on corehq/apps/app_manager/app_schemas/tests/test_case_properties.py:
```diff
@@ -32,6 +34,66 @@ def test_basic_apps(self):
+            'household_color',
+            'household_id',
+            'household_name',
+        ])
```
For these same tests to pass on master, I had to comment out the lines that test propagation of `parent/<property>` on the child up to `<property>` on the parent.

https://github.com/dimagi/commcare-hq/commit/ebefb5041faca1cc935a431b1be012045815afa1#diff-bc920545fa7fcaa4700d89bb675170caR75

dannyroberts commented on corehq/apps/app_manager/models.py:
```diff
@@ -1275,15 +1276,15 @@ def case_list_modules(self):
         return bool(self.case_list_modules)
 
-    def get_save_to_case_updates(self, case_type):
+    def get_save_to_case_updates(self):
```
You'll see a pattern in this file of taking methods that used to give all of the properties for a single (specified) case type and changing them to return properties for every case type they stumble upon in the process (as `{<case_type>: set([<property>])}`). This prevents us from having to know all of the case_types in advance and then ask each of these functions to do it for each of the case_types we know about, which cuts down on the number of times we have to execute these loops, but also (and more importantly) makes it easier to reason with.

dannyroberts committed:
rename get_case_relationships (function) to get_parent_type_map

I reused that name (get_case_relationships) for a method on ParentCasePropertyBuilder
which has a different return type, so now it is confusing. Using the same name as
the method that backs it, for consistency.

dannyroberts commented:
Thanks for the quick look at early versions of this @millerdev. Removed the don't-merge label, and am not planning to add more at this time, so this PR is now ready to review.

czue commented:
Wow this looks fantastic! :clap: :clap: 

I scanned the code but didn't review in depth since it seemed that others might have already done and be better suited for that. Happy to do a deep review if no one else jumps in though. I'm "approved" from a conceptual / high-level approach.

I've been thinking a lot about this problem and I'm just going to dump a few high-level things I've been thinking here as a comment (there's probably a more appropriate place for this but am not sure where that would be).

---

**1. I believe it would be very valuable to define a small number of interfaces that accomplish this task**

At a minimum I think we need one interface that generates a schema from apps and builds, and a second interface that generates it from "everything" (where everything might include apps, builds, and whatever other data models exist to serve this purpose - e.g. DD and export schemas).

You need the apps one to prepopulate the DD and exports, at a minimum, and possibly as a subset of the "everything" one.

**2. I believe we should take effort to create a consistent Directed hopefully-Acyclic Graph that defines the pipeline of how the "everything" API is produced.**

For example, one such graph might be something like:

```
Apps -- > DD --> export schemas ---> "everything else"
```

In my head the most confusing parts of creating this graph is deciding 1) where the DD lives and especially its relationship with Apps if they are not codependent, and 2) what is the last thing that powers "everything else" (I've expressed in a few other places that I think it should be export schemas, and that possibly export schemas should *become* the DD, though that's longer term thinking for sure).

**3. Anything that deals with case properties should (eventually) also always deal with datatypes**

As we move to a world where we try and provide better tools and checks around more concrete data models, I think proper data types will become more and more important. Granted, this is a very hard problem - particularly because it's not a given that a property even *has* a single type once you look across apps/builds. But I think it would benefit us to start thinking and designing our APIs in a way that could support this. (Note: at a minimum this *has* to be true about the "everything" API).

---

Anyway, all that said I believe this PR is a big step in a great direction, and anything that cleans and simplifies these interfaces is a good step forward. 

cc @proteusvacuum since we've also been chatting about this a bit.

kaapstorm commented on corehq/apps/app_manager/models.py:
```diff
@@ -1794,16 +1795,37 @@ def extended_build_validation(self, error_meta, xml_valid, validate_module=True)
+                format_key(*item) for item in self.actions.usercase_update.update.items()}
+        }
+
+    def get_all_case_updates(self):
```
nit: Could this move to `IndexedFormBase`? The methods this calls are different for `Form` and `AdvancedForm` but this method is identical.

kaapstorm commented:
This is great. I really like the visibility into where things are coming from that's made obvious by `get_all_case_updates()`.

dannyroberts committed:
push get_all_case_updates into IndexedFormBase

it was identical in all subclasses

dannyroberts committed:
push get_case_updates_for_case_type into IndexedFormBase

it was identical in all subclasses

millerdev commented on corehq/apps/app_manager/app_schemas/case_properties.py:
```diff
@@ -155,32 +429,39 @@ def get_parent_type_map(self, case_types, allow_multiple_parents=False):
-                        include_parent_properties=True):
+def get_parent_type_map(app):
+    builder = ParentCasePropertyBuilder(app)
+    return builder.get_parent_type_map(app.get_case_types())
```
What do you think of passing `allow_multiple_parents=True` here? or maybe at least expose that parameter so it can be passed if needed. Otherwise this will return incomplete and I think even nondeterministic results for some apps.

dannyroberts replied:
Okay, I changed the name of this arg, made the good value the default one, and made the one place that uses the bad value pass that in explicitly. Instead of `allow_multiple_parents=False` (which was the default) you now have to pass in `if_multiple_parents_arbitrarily_pick_one=True` explicitly.

millerdev commented on corehq/apps/app_manager/models.py:
```diff
@@ -1275,15 +1276,15 @@ def case_list_modules(self):
-                updates |= set(save_to_case_update.properties)
-        return updates
+            case_type = save_to_case_update.case_type
+            updates_by_case_type[case_type] |= set(save_to_case_update.properties)
```
nit: I think it's marginally more efficient to do
```py
updates_by_case_type[case_type].update(save_to_case_update.properties)
```
since that does not create a new set (and throw away the old one).

millerdev reviewed (approved):
This is great! Definitely makes it easier to understand case properties and relationships. Added some comments, I think none are blocking except maybe the one on `get_parent_type_map`.

dannyroberts committed:
use .update instead of |= set

dannyroberts committed:
name deprecated get_parent_type_map arg more jarringly

allow_multiple_parents => if_multiple_parents_arbitrarily_pick_one (with opposite value)
good behavior is now the default

dannyroberts commented on corehq/apps/app_manager/app_schemas/case_properties.py:
```diff
@@ -18,130 +20,402 @@
+        Unlike `get_case_relationships`, it doesn't include the `relationship`.
+        That is because it is almost always "parent". Presumably, if this method is used
+        without care, it will cause bugs in the rarer case when relationship is *not* "parent".
+        # todo audit usage to see if such bugs can occur
```
I don't want to mess with that actually because I think the one place that does use this may actually use it correctly? As in, it just uses it to figure out what all of the other "relevant" case types are, but doesn't actually assume that they have relationship parent

millerdev replied:
👍 that makes sense

dannyroberts committed:
fix some docstrings

dannyroberts committed:
optimize ParentCasePropertyBuilder data dictionary case property query

dannyroberts committed:
avoid creating intermediary set when generator will do

czue commented:
> I'm sure that'll be nice to come back to when someone starts working on it!

:+1: 

what I do think might be important - and am not quite sure how to achieve - is that we get the relevant parties aligned on what we want the shape of the future state of the world to be and we make sure that gets communicated *before* someone starts working on it. else i think people will be prone to implement things in the way that they figure out first and we'll continue to build a fractured and confusing state of affairs instead of working towards a more cohesive future state. and either it will slip through review because no-one knows/agrees on the vision, or because we did know the vision but we felt bad asking people to redo work. any thoughts you have on how better to achieve consensus and communicate that out appropriately without overwhelming people would be welcome. i'd be happy to draft a more formal doc if you think that would help (and if you agree with the note i wrote above)

kaapstorm commented on corehq/apps/app_manager/app_schemas/case_properties.py:
```diff
@@ -393,17 +393,17 @@ def get_properties_by_case_type(self):
         )
 
-    def get_parent_type_map(self, case_types, allow_multiple_parents=False):
+    def get_parent_type_map(self, case_types, if_multiple_parents_arbitrarily_pick_one=False):
```
:upside_down_face: :black_heart:
